File: tui/src/api/proxmox.zig
const std = @import("std");
const config = @import("../config.zig");
const http = @import("http_client.zig");
const Allocator = std.mem.Allocator;

pub const VmStatus = struct {
    vmid: i64,
    name: []const u8,
    status: []const u8,
    node: []const u8,
    maxdisk: i64,
};

pub const StoragePool = struct {
    name: []const u8,
    node: []const u8,
    pool_type: []const u8,
    status: []const u8,
    disk: i64,
    maxdisk: i64,
};

pub const NodeStatus = struct {
    node: []const u8,
    status: []const u8,
    cpu: f64,
    mem: i64,
    maxmem: i64,
    uptime: i64,
};

pub const BackupEntry = struct {
    volid: []const u8,
    node: []const u8,
    storage: []const u8,
    size: i64,
    ctime: i64,
    vmid: i64,
    format: []const u8,
};

pub const ProxmoxClient = struct {
    client: http.HttpClient,
    allocator: Allocator,

    pub fn init(allocator: Allocator, pve: config.ProxmoxCluster) ProxmoxClient {
        return .{
            .client = http.HttpClient.init(allocator, pve),
            .allocator = allocator,
        };
    }

    /// Fetch all VM resources across the PVE cluster.
    pub fn getClusterResources(self: *ProxmoxClient) ![]VmStatus {
        const body = self.client.get("/api2/json/cluster/resources?type=vm") catch {
            return &.{};
        };
        defer self.allocator.free(body);

        var parsed = http.parseJsonResponse(self.allocator, body) catch {
            return &.{};
        };
        defer parsed.deinit();

        const data_val = switch (parsed.value) {
            .object => |obj| obj.get("data") orelse return &.{},
            else => return &.{},
        };
        const items = switch (data_val) {
            .array => |arr| arr.items,
            else => return &.{},
        };

        var results: std.ArrayListUnmanaged(VmStatus) = .empty;
        for (items) |item| {
            const obj = switch (item) {
                .object => |o| o,
                else => continue,
            };

            // Only include QEMU VMs (not LXC containers)
            const res_type = http.jsonStr(obj, "type", "");
            if (!std.mem.eql(u8, res_type, "qemu")) continue;

            const name = try self.allocator.dupe(u8, http.jsonStr(obj, "name", "unknown"));
            const status = try self.allocator.dupe(u8, http.jsonStr(obj, "status", "unknown"));
            const node = try self.allocator.dupe(u8, http.jsonStr(obj, "node", "unknown"));

            try results.append(self.allocator, .{
                .vmid = http.jsonInt(obj, "vmid", 0),
                .name = name,
                .status = status,
                .node = node,
                .maxdisk = http.jsonInt(obj, "maxdisk", 0),
            });
        }

        return results.toOwnedSlice(self.allocator);
    }

    /// Fetch status for a specific PVE node.
    pub fn getNodeStatus(self: *ProxmoxClient, node: []const u8) !?NodeStatus {
        const path = try std.fmt.allocPrint(self.allocator, "/api2/json/nodes/{s}/status", .{node});
        defer self.allocator.free(path);

        const body = self.client.get(path) catch return null;
        defer self.allocator.free(body);

        var parsed = http.parseJsonResponse(self.allocator, body) catch return null;
        defer parsed.deinit();

        const data_val = switch (parsed.value) {
            .object => |obj| obj.get("data") orelse return null,
            else => return null,
        };
        const obj = switch (data_val) {
            .object => |o| o,
            else => return null,
        };

        return .{
            .node = try self.allocator.dupe(u8, node),
            .status = try self.allocator.dupe(u8, http.jsonStr(obj, "status", "unknown")),
            .cpu = http.jsonFloat(obj, "cpu", 0),
            .mem = http.jsonInt(obj, "mem", 0),
            .maxmem = http.jsonInt(obj, "maxmem", 0),
            .uptime = http.jsonInt(obj, "uptime", 0),
        };
    }

    /// Fetch all storage pools across the PVE cluster.
    pub fn getStoragePools(self: *ProxmoxClient) ![]StoragePool {
        const body = self.client.get("/api2/json/cluster/resources?type=storage") catch {
            return &.{};
        };
        defer self.allocator.free(body);

        var parsed = http.parseJsonResponse(self.allocator, body) catch {
            return &.{};
        };
        defer parsed.deinit();

        const data_val = switch (parsed.value) {
            .object => |obj| obj.get("data") orelse return &.{},
            else => return &.{},
        };
        const items = switch (data_val) {
            .array => |arr| arr.items,
            else => return &.{},
        };

        var results: std.ArrayListUnmanaged(StoragePool) = .empty;
        for (items) |item| {
            const obj = switch (item) {
                .object => |o| o,
                else => continue,
            };

            const name = try self.allocator.dupe(u8, http.jsonStr(obj, "storage", "unknown"));
            const node = try self.allocator.dupe(u8, http.jsonStr(obj, "node", "unknown"));
            const pool_type = try self.allocator.dupe(u8, http.jsonStr(obj, "plugintype", http.jsonStr(obj, "type", "unknown")));
            const status = try self.allocator.dupe(u8, http.jsonStr(obj, "status", "unknown"));

            try results.append(self.allocator, .{
                .name = name,
                .node = node,
                .pool_type = pool_type,
                .status = status,
                .disk = http.jsonInt(obj, "disk", 0),
                .maxdisk = http.jsonInt(obj, "maxdisk", 0),
            });
        }

        return results.toOwnedSlice(self.allocator);
    }

    /// List vzdump backups from a specific storage pool on a node.
    pub fn listBackups(self: *ProxmoxClient, node: []const u8, storage: []const u8) ![]BackupEntry {
        const path = try std.fmt.allocPrint(self.allocator, "/api2/json/nodes/{s}/storage/{s}/content?content=backup", .{ node, storage });
        defer self.allocator.free(path);

        const body = self.client.get(path) catch return &.{};
        defer self.allocator.free(body);

        var parsed = http.parseJsonResponse(self.allocator, body) catch return &.{};
        defer parsed.deinit();

        const data_val = switch (parsed.value) {
            .object => |obj| obj.get("data") orelse return &.{},
            else => return &.{},
        };
        const items = switch (data_val) {
            .array => |arr| arr.items,
            else => return &.{},
        };

        var results: std.ArrayListUnmanaged(BackupEntry) = .empty;
        for (items) |item| {
            const obj = switch (item) {
                .object => |o| o,
                else => continue,
            };

            const volid = try self.allocator.dupe(u8, http.jsonStr(obj, "volid", ""));
            const format = try self.allocator.dupe(u8, http.jsonStr(obj, "format", "unknown"));

            try results.append(self.allocator, .{
                .volid = volid,
                .node = try self.allocator.dupe(u8, node),
                .storage = try self.allocator.dupe(u8, storage),
                .size = http.jsonInt(obj, "size", 0),
                .ctime = http.jsonInt(obj, "ctime", 0),
                .vmid = http.jsonInt(obj, "vmid", 0),
                .format = format,
            });
        }

        return results.toOwnedSlice(self.allocator);
    }

    /// Delete a backup by volume ID.
    pub fn deleteBackup(self: *ProxmoxClient, node: []const u8, storage: []const u8, volid: []const u8) !void {
        // URL-encode the volid (colons → %3A)
        var encoded: std.ArrayListUnmanaged(u8) = .empty;
        defer encoded.deinit(self.allocator);
        for (volid) |c| {
            if (c == ':') {
                try encoded.appendSlice(self.allocator, "%3A");
            } else {
                try encoded.append(self.allocator, c);
            }
        }

        const path = try std.fmt.allocPrint(self.allocator, "/api2/json/nodes/{s}/storage/{s}/content/{s}", .{
            node, storage, encoded.items,
        });
        defer self.allocator.free(path);

        const body = try self.client.delete(path);
        self.allocator.free(body);
    }

    pub fn deinit(self: *ProxmoxClient) void {
        self.client.deinit();
    }
};

File: tui/src/integrations/proxmox.rs
use crate::config::ProxmoxCluster;
use crate::integrations::command;
use anyhow::{Context, Result};
use serde_json::Value;

#[derive(Debug, Clone, Default)]
pub struct VmStatus {
    pub vmid: i64,
    pub name: String,
    pub status: String,
    pub node: String,
    pub maxdisk: i64,
}

#[derive(Debug, Clone, Default)]
pub struct StoragePool {
    pub name: String,
    pub node: String,
    pub pool_type: String,
    pub status: String,
    pub disk: i64,
    pub maxdisk: i64,
}

#[derive(Debug, Clone, Default)]
pub struct NodeStatus {
    pub node: String,
    pub cpu: f64,
    pub mem: i64,
    pub maxmem: i64,
}

#[derive(Debug, Clone, Default)]
pub struct BackupEntry {
    pub volid: String,
    pub node: String,
    pub storage: String,
    pub size: i64,
    pub ctime: i64,
    pub vmid: i64,
}

pub fn get_cluster_resources(cluster: &ProxmoxCluster) -> Result<Vec<VmStatus>> {
    let body = request(cluster, "GET", "/api2/json/cluster/resources?type=vm")?;
    let items = data_array(&body)?;
    Ok(items
        .into_iter()
        .filter(|item| item.get("type").and_then(Value::as_str) == Some("qemu"))
        .map(|item| VmStatus {
            vmid: json_i64(&item, "vmid"),
            name: json_string(&item, "name", "unknown"),
            status: json_string(&item, "status", "unknown"),
            node: json_string(&item, "node", "unknown"),
            maxdisk: json_i64(&item, "maxdisk"),
        })
        .collect())
}

pub fn get_storage_pools(cluster: &ProxmoxCluster) -> Result<Vec<StoragePool>> {
    let body = request(cluster, "GET", "/api2/json/cluster/resources?type=storage")?;
    let items = data_array(&body)?;
    Ok(items
        .into_iter()
        .map(|item| StoragePool {
            name: json_string(&item, "storage", "unknown"),
            node: json_string(&item, "node", "unknown"),
            pool_type: item
                .get("plugintype")
                .and_then(Value::as_str)
                .or_else(|| item.get("type").and_then(Value::as_str))
                .unwrap_or("unknown")
                .to_string(),
            status: json_string(&item, "status", "unknown"),
            disk: json_i64(&item, "disk"),
            maxdisk: json_i64(&item, "maxdisk"),
        })
        .collect())
}

pub fn get_node_status(cluster: &ProxmoxCluster, node: &str) -> Result<NodeStatus> {
    let body = request(cluster, "GET", &format!("/api2/json/nodes/{node}/status"))?;
    let data = data_object(&body)?;
    Ok(NodeStatus {
        node: node.to_string(),
        cpu: json_f64(&data, "cpu"),
        mem: json_i64(&data, "mem"),
        maxmem: json_i64(&data, "maxmem"),
    })
}

pub fn list_backups(
    cluster: &ProxmoxCluster,
    node: &str,
    storage: &str,
) -> Result<Vec<BackupEntry>> {
    let body = request(
        cluster,
        "GET",
        &format!("/api2/json/nodes/{node}/storage/{storage}/content?content=backup"),
    )?;
    let items = data_array(&body)?;
    Ok(items
        .into_iter()
        .map(|item| BackupEntry {
            volid: json_string(&item, "volid", ""),
            node: node.to_string(),
            storage: storage.to_string(),
            size: json_i64(&item, "size"),
            ctime: json_i64(&item, "ctime"),
            vmid: json_i64(&item, "vmid"),
        })
        .collect())
}

pub fn delete_backup(
    cluster: &ProxmoxCluster,
    node: &str,
    storage: &str,
    volid: &str,
) -> Result<()> {
    let encoded = volid.replace(':', "%3A").replace('/', "%2F");
    let _ = request(
        cluster,
        "DELETE",
        &format!("/api2/json/nodes/{node}/storage/{storage}/content/{encoded}"),
    )?;
    Ok(())
}

fn request(cluster: &ProxmoxCluster, method: &str, path: &str) -> Result<String> {
    let curl = command::resolve_binary("curl", "PVT_CURL_BIN")?;
    let url = format!("{}{}", cluster.endpoint, path);
    let mut argv = vec![
        curl,
        "--silent".to_string(),
        "--show-error".to_string(),
        "--fail".to_string(),
        "--max-time".to_string(),
        "10".to_string(),
        "--config".to_string(),
        "-".to_string(),
    ];
    if method != "GET" {
        argv.push("-X".to_string());
        argv.push(method.to_string());
    }
    if !cluster.tls_verify {
        argv.push("-k".to_string());
    }
    let curl_config = format!(
        "url = \"{}\"\nheader = \"Authorization: PVEAPIToken={}={}\"\n",
        escape_curl_config(&url),
        escape_curl_config(&cluster.token_id),
        escape_curl_config(&cluster.token_secret),
    );
    command::run_with_input(&argv, Some(&curl_config), 1024 * 1024)
        .context("proxmox request failed")
}

fn escape_curl_config(value: &str) -> String {
    value.replace('\\', "\\\\").replace('"', "\\\"")
}

fn data_array(body: &str) -> Result<Vec<Value>> {
    let parsed: Value = serde_json::from_str(body).context("invalid proxmox JSON")?;
    Ok(parsed
        .get("data")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default())
}

fn data_object(body: &str) -> Result<Value> {
    let parsed: Value = serde_json::from_str(body).context("invalid proxmox JSON")?;
    Ok(parsed.get("data").cloned().unwrap_or(Value::Null))
}

fn json_string(value: &Value, key: &str, default: &str) -> String {
    value
        .get(key)
        .and_then(Value::as_str)
        .unwrap_or(default)
        .to_string()
}

fn json_i64(value: &Value, key: &str) -> i64 {
    value
        .get(key)
        .and_then(|value| value.as_i64().or_else(|| value.as_f64().map(|v| v as i64)))
        .unwrap_or_default()
}

fn json_f64(value: &Value, key: &str) -> f64 {
    value
        .get(key)
        .and_then(|value| value.as_f64().or_else(|| value.as_i64().map(|v| v as f64)))
        .unwrap_or_default()
}
